### DISCUSSION TITLE
posts/fork

### DISCUSSION THREAD

**Discussion by giscus** [whexy/whexy-blog-comments#7] [General]
# posts/fork/

a student obsessed with the computing world

https://whexy.com/posts/fork/

**Comment by MstMoonshine** [comment]
Two questions:
(1). When vfork is not available (for example, when vfork is not safe), is posix_spawn() still faster compared to fork()?  
(2). How did you guarantee the safety of vfork in the fuzzing machine?

**Reply by whexy** [reply]
Linux used to allow setting `POSIX_SPAWN_USEVFORK` flag to force using `vfork()` instead of `fork()`. Since glibc 2.24, this flag has no effect. Linux decides to get smarter and it will make decision automatically for the user.

> When vfork is not available (for example, when vfork is not safe), is posix_spawn() still faster compared to fork()?

No, it will fall back to `fork(); exec();`. So it's almost the same.

> How did you guarantee the safety of vfork in the fuzzing machine?

We cannot control the safety of `vfork`. But at least we can create a relatively safer environment. For example, avoid using multiple threads when invoking `vfork()`. Unsafe condition is not commonly seen.
